### PR TITLE
QUICKSTEP-20: Supported Window Aggregation for AVG

### PR DESCRIPTION
**Pull request by shixuan-fan**
Quickstep could now make the calculation for window aggregation function average. The work-flow is:

`ExecutionGenerator` constructs `WindowAggregationOperator` and initializes `WindowAggregationOperationState`
=> `WindowAggregationOperator::execute()` is called
=> `WindowAggregationOperationState::windowAggregate()`: combines all blocks' tuples into one `ValueAccessor`
=> `WindowAggregationHandle::calculate()`: actually does the calculation and return a `ColumnVector*` including all the calculated window aggregates
=> `WindowAggregationOperationState::windowAggregate()` will add the new `ColumnVector` into the created `ValueAccessor` and output the result to the `output_destination`.

We are using the simplest approach for the calculation: one thread will iterate through all the tuples. For each tuple, the thread will find all the elements in the given window and calculate the result.

In current version, we don't support `RANGE` frame mode. Also, other functions' handle will be created in the next PR.

### CODE REVIEW

File: expressions/window_aggregation/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+add_library(quickstep_expressions_windowaggregation_WindowAggregateFunctionMax
+            WindowAggregateFunctionMax.cpp
+            WindowAggregateFunctionMax.hpp)
+add_library(quickstep_expressions_windowaggregation_WindowAggregateFunctionMin
+            WindowAggregateFunctionMin.cpp
+            WindowAggregateFunctionMin.hpp)
+add_library(quickstep_expressions_windowaggregation_WindowAggregateFunctionSum
+            WindowAggregateFunctionSum.cpp
+            WindowAggregateFunctionSum.hpp)
+add_library(quickstep_expressions_windowaggregation_WindowAggregateFunctionFactory
```
Comment (zuyu) [inline expressions/window_aggregation/CMakeLists.txt]: Code style: please move it before `Max`.

Comment (shixuan-fan) [inline expressions/window_aggregation/CMakeLists.txt]: Fixed! :)

File: expressions/window_aggregation/WindowAggregateFunction.hpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+#include "expressions/window_aggregation/WindowAggregateFunction.pb.h"
+#include "expressions/window_aggregation/WindowAggregationID.hpp"
+#include "storage/StorageBlockInfo.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+class CatalogRelationSchema;
+class WindowAggregationHandle;
+class Type;
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunction.hpp]: Code style: Reorder declarations.

Comment (shixuan-fan) [inline expressions/window_aggregation/WindowAggregateFunction.hpp]: Done!

File: expressions/window_aggregation/CMakeLists.txt (hunk 1 above)
Comment (zuyu) [inline expressions/window_aggregation/CMakeLists.txt]: Unfortunately, I did not see the proposed changes.

File: expressions/window_aggregation/WindowAggregateFunctionAvg.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+bool WindowAggregateFunctionAvg::canApplyToTypes(
+    const std::vector<const Type*> &argument_types) const {
+  // AVG is unary.
+  if (argument_types.size() != 1) {
+    return false;
+  }
+
+  // Argument must be addable and divisible.
+  return BinaryOperationFactory::GetBinaryOperation(BinaryOperationID::kAdd)
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionAvg.cpp]: [Code style](https://google.github.io/styleguide/cppguide.html#Return_Values):

Please refactor to

```
return xxx &&
           yyy;
```

File: expressions/window_aggregation/WindowAggregateFunctionCount.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    const std::vector<const Type*> &argument_types) const {
+  if (!canApplyToTypes(argument_types)) {
+    return nullptr;
+  }
+
+  return &TypeFactory::GetType(kLong);
+}
+
+WindowAggregationHandle* WindowAggregateFunctionCount::createHandle(
+    std::vector<const Type*> &&argument_types,
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionCount.cpp]: I am curious why we use the rvalue here?

File: expressions/window_aggregation/WindowAggregateFunctionCount.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+
+#include "expressions/window_aggregation/WindowAggregateFunction.hpp"
+#include "expressions/window_aggregation/WindowAggregationID.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+class WindowAggregationHandle;
+class Type;
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionCount.hpp]: Code style: Declaration ordering.

File: expressions/window_aggregation/WindowAggregateFunctionFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  } else if (name == "sum") {
+    return &WindowAggregateFunctionSum::Instance();
+  } else {
+    return nullptr;
+  }
+}
+
+bool WindowAggregateFunctionFactory::ProtoIsValid(
+    const serialization::WindowAggregateFunction &proto) {
+  return proto.IsInitialized()
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionFactory.cpp]: [Code style](https://google.github.io/styleguide/cppguide.html#Return_Values):

Please refactor to
```
  return xxx &&
         yyy;
```

File: expressions/window_aggregation/WindowAggregateFunctionFactory.hpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+ *        implementations of WindowAggregateFunction.
+ *
+ * WindowAggregateFunctionFactory allows client code to use any
+ * WindowAggregateFunction in Quickstep in a generic way without having to know
+ * about all the specific subclasses of WindowAggregateFunction. In particular,
+ * it is used to deserialize WindowAggregateFunctions used in
+ * WindowAggregationOperationState from their protobuf representations
+ * (originally created by the optimizer) when deserializing a QueryContext.
+ **/
+class WindowAggregateFunctionFactory {
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionFactory.hpp]: It is optional, and we could do the refactor later according to this [code style](https://google.github.io/styleguide/cppguide.html#Nonmember,_Static_Member,_and_Global_Functions).

We have to use the namespace install of static member functions.

File: expressions/window_aggregation/WindowAggregateFunctionMax.hpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+
+#include "expressions/window_aggregation/WindowAggregateFunction.hpp"
+#include "expressions/window_aggregation/WindowAggregationID.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+class WindowAggregationHandle;
+class Type;
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionMax.hpp]: Code style: reorder the declaration.

File: expressions/window_aggregation/WindowAggregateFunctionMin.hpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+
+#include "expressions/window_aggregation/WindowAggregateFunction.hpp"
+#include "expressions/window_aggregation/WindowAggregationID.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+class WindowAggregationHandle;
+class Type;
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionMin.hpp]: Code style: reorder the declaration.

File: expressions/window_aggregation/WindowAggregateFunctionSum.hpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+
+#include "expressions/window_aggregation/WindowAggregateFunction.hpp"
+#include "expressions/window_aggregation/WindowAggregationID.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+class WindowAggregationHandle;
+class Type;
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregateFunctionSum.hpp]: Code style: reorder the declaration.

File: expressions/window_aggregation/WindowAggregationHandle.hpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                                  const std::int64_t num_following) const = 0;
+
+ protected:
+  /**
+   * @brief Constructor.
+   *
+   * @param partition_key_types The Types of the partition key.
+   **/
+  explicit WindowAggregationHandle(
+      std::vector<const Type*> &&partition_key_types) {
```
Comment (zuyu) [inline expressions/window_aggregation/WindowAggregationHandle.hpp]: I was wondering why using a rvalue here.

File: expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+        TypeFactory::GetType(typeID, static_cast<std::size_t>(10)) :
+        TypeFactory::GetType(typeID);
+
+    return WindowAggregateFunctionFactory::Get(WindowAggregationID::kAvg).canApplyToTypes(
+        std::vector<const Type*>(1, &type));
+  }
+
+  // Test resultTypeForArgumentTypes().
+  static bool ResultTypeForArgumentTypesTest(TypeID input_type_id,
+                                            TypeID output_type_id) {
```
Comment (zuyu) [inline expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp]: Code style: add one whitespace indentation.

File: expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+    const GenericType &type = GenericType::Instance(true);
+    initializeHandle(type);
+
+    // Create argument, partition key and cpptype vectors.
+    std::vector<typename GenericType::cpptype*> argument_cpp_vector;
+    argument_cpp_vector.reserve(kNumTuples);
+    ColumnVector *argument_type_vector =
+        createArgumentGeneric<GenericType>(&argument_cpp_vector);
+    NativeColumnVector *partition_key_vector =
+      new NativeColumnVector(IntType::InstanceNonNullable(), kNumTuples + 2);
```
Comment (zuyu) [inline expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp]: Code style: add two more whitespace indentation.

File: expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+
+      typename GenericType::cpptype sum;
+      SetDataType(0, &sum);
+      sum += *argument_cpp_vector[i];
+      int count = 1;
+      bool is_null = false;
+
+      for (std::size_t precede = 1; precede <= kNumPreceding; ++precede) {
+        // Not the same partition.
+        if (i / kNumTuplesPerPartition != (i - precede) / kNumTuplesPerPartition
```
Comment (zuyu) [inline expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp]: [Code style](https://google.github.io/styleguide/cppguide.html#Boolean_Expressions):

Please refactor to
```
  if (xxx ||
      yyy)
```

File: expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+          break;
+        }
+
+        sum += *argument_cpp_vector[i - precede];
+        count++;
+      }
+
+      for (int follow = 1; follow <= kNumPreceding; ++follow) {
+        // Not the same partition.
+        if (i / kNumTuplesPerPartition != (i + follow) / kNumTuplesPerPartition
```
Comment (zuyu) [inline expressions/window_aggregation/tests/WindowAggregationHandleAvg_unittest.cpp]: Ditto for the code style for boolean expressions:

https://google.github.io/styleguide/cppguide.html#Boolean_Expressions

File: query_optimizer/expressions/WindowAggregateFunction.hpp [hunk 16]
```diff
@@ -33,7 +33,7 @@
 
 namespace quickstep {
 
-class AggregateFunction;
+class WindowAggregateFunction;
```
Comment (zuyu) [inline query_optimizer/expressions/WindowAggregateFunction.hpp]: Code style: reorder the declaration.

File: query_optimizer/resolver/Resolver.cpp [hunk 17]
```diff
@@ -2656,11 +2666,12 @@ E::ScalarPtr Resolver::resolveFunctionCall(
   }
 
   // Make sure a naked COUNT() with no arguments wasn't specified.
-  if ((aggregate->getAggregationID() == AggregationID::kCount)
-      && (resolved_arguments.empty())
-      && (!count_star)) {
-    THROW_SQL_ERROR_AT(&parse_function_call)
-        << "COUNT aggregate requires an argument (either scalar or star (*))";
+  if ((aggregate != nullptr && aggregate->getAggregationID() == AggregationID::kCount)
```
Comment (zuyu) [inline query_optimizer/resolver/Resolver.cpp]: Code style for the bool expressions:

https://google.github.io/styleguide/cppguide.html#Boolean_Expressions

File: query_optimizer/resolver/Resolver.cpp [hunk 18]
```diff
@@ -2670,7 +2681,8 @@ E::ScalarPtr Resolver::resolveFunctionCall(
   }
 
   // Make sure that the aggregate can apply to the specified argument(s).
-  if (!aggregate->canApplyToTypes(argument_types)) {
+  if ((aggregate != nullptr && !aggregate->canApplyToTypes(argument_types))
```
Comment (zuyu) [inline query_optimizer/resolver/Resolver.cpp]: Code style for the bool expressions:

https://google.github.io/styleguide/cppguide.html#Boolean_Expressions

File: query_optimizer/tests/execution_generator/Select.test [hunk 19]
```diff
@@ -953,19 +953,79 @@ WHERE double_col < 0
 ==
 
 # Window Aggregation Test.
-# Currently this is not supported, an empty table will be returned.
-SELECT avg(int_col) OVER w FROM test
+SELECT char_col, long_col, avg(long_col) OVER w FROM test
 WINDOW w AS
-(PARTITION BY char_col
- ORDER BY long_col DESC NULLS LAST
+(ORDER BY char_col DESC NULLS LAST
```
Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: Quick question: What if we remove the `order by` clause?

File: relational_operators/WindowAggregationOperator.cpp [hunk 20]
```diff
@@ -67,6 +73,11 @@ serialization::WorkOrder* WindowAggregationOperator::createWorkOrderProto() {
   proto->set_query_id(query_id_);
   proto->SetExtension(serialization::WindowAggregationWorkOrder::window_aggr_state_index,
                       window_aggregation_state_index_);
+
+  std::vector<block_id> relation_blocks = input_relation_.getBlocksSnapshot();
```
Comment (zuyu) [inline relational_operators/WindowAggregationOperator.cpp]: We could do some improvements on the code:

```
  const std::vector<block_id> relation_blocks = input_relation_.getBlocksSnapshot();
  for (const block_id bid : relation_blocks) { ... }
```

File: relational_operators/WindowAggregationOperator.hpp [hunk 21]
```diff
@@ -58,16 +58,19 @@ class WindowAggregationOperator : public RelationalOperator {
...
    * @param query_id The ID of this query.
    * @param input_relation The relation to perform aggregation over.
+   * @param output_relation The relation for output.
    * @param window_aggregation_state_index The index of WindowAggregationState
    *                                       in QueryContext.
    * @param output_destination_index The index of InsertDestination in
    *                                 QueryContext for the output.
    **/
   WindowAggregationOperator(const std::size_t query_id,
+                            const CatalogRelation &input_relation,
```
Comment (zuyu) [inline relational_operators/WindowAggregationOperator.hpp]: Remove this line?

File: relational_operators/WorkOrder.proto [hunk 22]
```diff
@@ -249,6 +249,7 @@ message WindowAggregationWorkOrder {
   extend WorkOrder {
     // All required
     optional uint32 window_aggr_state_index = 336;
-    optional int32 insert_destination_index = 337;
+    repeated fixed64 block_ids = 337;
```
Comment (zuyu) [inline relational_operators/WorkOrder.proto]: FYI, we don't need to change the number assigned to `insert_destination_index`, but just add a new field for `block_ids` while assigning `338`.

File: storage/WindowAggregationOperationState.cpp [hunk 23]
```diff
@@ -55,14 +62,12 @@ WindowAggregationOperationState::WindowAggregationOperationState(
     StorageManager *storage_manager)
     : input_relation_(input_relation),
       arguments_(std::move(arguments)),
-      partition_by_attributes_(std::move(partition_by_attributes)),
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: Question: are we removing the support for the partitioned tuples?

File: storage/WindowAggregationOperationState.cpp [hunk 24]
```diff
@@ -176,4 +171,122 @@ bool WindowAggregationOperationState::ProtoIsValid(const serialization::WindowAg
...
+  // work. This will need Order Key to be passed so that we know where the
+  // window should start and end.
+  if (!is_row_) {
+    std::cout << "Currently we don't support RANGE frame mode :(\n";
+    return;
+  }
+
+  // Get the total number of tuples.
+  int num_tuples = 0;
+  for (block_id block_idx : block_ids) {
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: Add `const` for `block_id block_idx`.

File: storage/WindowAggregationOperationState.cpp [hunk 25]
```diff
@@ -176,4 +171,122 @@ bool WindowAggregationOperationState::ProtoIsValid(const serialization::WindowAg
...
+    if (NativeColumnVector::UsableForType(type)) {
+      attribute_vecs.push_back(new NativeColumnVector(type, num_tuples));
+    } else {
+      attribute_vecs.push_back(new IndirectColumnVector(type, num_tuples));
+    }
+  }
+
+  // Construct column vectors for arguments.
+  std::vector<ColumnVector*> argument_vecs;
+  for (std::unique_ptr<const Scalar> &argument : arguments_) {
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: Add `const` for `std::unique_ptr<const Scalar> &argument`.

File: storage/WindowAggregationOperationState.cpp [hunk 26]
```diff
@@ -176,4 +171,122 @@ bool WindowAggregationOperationState::ProtoIsValid(const serialization::WindowAg
...
+    if (NativeColumnVector::UsableForType(type)) {
+      argument_vecs.push_back(new NativeColumnVector(type, num_tuples));
+    } else {
+      argument_vecs.push_back(new IndirectColumnVector(type, num_tuples));
+    }
+  }
+
+  // TODO(Shixuan): Add Support for Vector Copy Elision Selection.
+  // Add tuples and arguments into ColumnVectors.
+  for (block_id block_idx : block_ids) {
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: Add `const` for `block_idx`.

File: storage/WindowAggregationOperationState.cpp [hunk 27]
```diff
@@ -176,4 +171,122 @@ bool WindowAggregationOperationState::ProtoIsValid(const serialization::WindowAg
...
+    BlockReference block = storage_manager_->getBlock(block_idx, input_relation_);
+    const TupleStorageSubBlock &tuple_block = block->getTupleStorageSubBlock();
+    SubBlocksReference sub_block_ref(tuple_block,
+                                     block->getIndices(),
+                                     block->getIndicesConsistent());
+    ValueAccessor *tuple_accessor = tuple_block.createValueAccessor();
+    ColumnVectorsValueAccessor *argument_accessor = new ColumnVectorsValueAccessor();
+    for (std::unique_ptr<const Scalar> &argument : arguments_) {
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: Add `const` for `argument`.

File: storage/WindowAggregationOperationState.proto [hunk 28]
```diff
@@ -19,15 +19,15 @@ syntax = "proto2";
...
 message WindowAggregationOperationState {
-  required int32 relation_id = 1;
-  required AggregateFunction function = 2;
-  repeated Scalar arguments = 3;
-  repeated Scalar partition_by_attributes = 4;
-  required bool is_row = 5;
-  required int64 num_preceding = 6;  // -1 means UNBOUNDED PRECEDING.
-  required int64 num_following = 7;  // -1 means UNBOUNDED FOLLOWING.
+  required int32 input_relation_id = 1;
+  required WindowAggregateFunction function = 3;
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.proto]: Something wrong w/ the number assignments and below.

File: storage/WindowAggregationOperationState.cpp [hunk 29]
```diff
@@ -126,8 +121,8 @@ WindowAggregationOperationState* WindowAggregationOperationState::ReconstructFro
   const std::int64_t num_preceding = proto.num_preceding();
   const std::int64_t num_following = proto.num_following();
 
-  return new WindowAggregationOperationState(database.getRelationSchemaById(proto.relation_id()),
-                                             aggregate_function,
+  return new WindowAggregationOperationState(database.getRelationSchemaById(proto.input_relation_id()),
+                                             window_aggregate_function,
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.cpp]: FYI, we could directly use `&WindowAggregateFunctionFactory::ReconstructFromProto(proto.function())` for `window_aggregate_function`.

**Comment by zuyu** [comment]
@shixuan-fan I've done one pass. Let me know if you want another.

File: expressions/window_aggregation/CMakeLists.txt (hunk 1 above)
Comment (shixuan-fan) [inline expressions/window_aggregation/CMakeLists.txt]: That's weird. Maybe I missed this commit during the squashing. Will fix it in the next commit.

File: expressions/window_aggregation/WindowAggregateFunctionCount.cpp (hunk 4 above)
Comment (shixuan-fan) [inline expressions/window_aggregation/WindowAggregateFunctionCount.cpp]: You are right. A reference would be better.

File: query_optimizer/tests/execution_generator/Select.test (hunk 19 above)
Comment (shixuan-fan) [inline query_optimizer/tests/execution_generator/Select.test]: Then the table will be in the same order as the input table when the window aggregates are calculated. And the calculated window aggregate for each row might depend on different values.

File: relational_operators/WindowAggregationOperator.cpp (hunk 20 above)
Comment (shixuan-fan) [inline relational_operators/WindowAggregationOperator.cpp]: Good catch!

File: storage/WindowAggregationOperationState.cpp (hunk 23 above)
Comment (shixuan-fan) [inline storage/WindowAggregationOperationState.cpp]: We will use attribute_id instead of Scalar.

File: storage/WindowAggregationOperationState.cpp (hunk 29 above)
Comment (shixuan-fan) [inline storage/WindowAggregationOperationState.cpp]: Sounds great.

**Comment by shixuan-fan** [comment]
@zuyu Thanks for all your efforts! I modified the codes according to your suggestion. Your second pass will be most appreciated. Take your time, no hurry.

File: relational_operators/WorkOrder.proto (hunk 22 above)
Comment (zuyu) [inline relational_operators/WorkOrder.proto]: So, do we want to refactor this changes?

File: query_optimizer/tests/execution_generator/Select.test (hunk 19 above)
Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: I get it, but do we have other alternatives to keep the semantics of `window aggregation`: aggregating over incoming tuples.

It particular makes more sense for the following next test cases w/ `NULL`s.

File: relational_operators/WorkOrder.proto (hunk 22 above)
Comment (shixuan-fan) [inline relational_operators/WorkOrder.proto]: Let's leave it as this to match the order in the fields of `WindowAggregationWorkOrder`.

File: query_optimizer/tests/execution_generator/Select.test (hunk 19 above)
Comment (shixuan-fan) [inline query_optimizer/tests/execution_generator/Select.test]: I don't quite understand. Could you please clarify this?

Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: I mean, we print the aggregation result according to the tuple order in the input table.

To that end, we may use a tuple id as the sort attribute.

Comment (shixuan-fan) [inline query_optimizer/tests/execution_generator/Select.test]: Ah I get what you are saying. In that case we don't actually need to do a sort since nothing will be changed, and the sort attribute will be empty (though now we don't pass the vector into the handle).

The sort attribute is not needed currently since we only support ROWS mode. To support RANGE mode, we have to add sort attribute to calculate the window frame. They will be introduced in the future PR.

Thanks for pointing that out :)

Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: So, any proposed changes?

Comment (shixuan-fan) [inline query_optimizer/tests/execution_generator/Select.test]: When the `RANGE` mode is introduced, a new parameter will be added to the `WindowAggregationHandle::calculate()` method, and would most likely be `const std::vector<attribute_id> &order_by_ids`. As for the cases where `ORDER BY` clause is missing, the vector would simply be empty, which indicates that the table is sorted by `tuple_id`. 

I don't think we need to explicitly add `tuple_id` as a sort attribute, because it is not actually an attribute.

Comment (shixuan-fan) [inline query_optimizer/tests/execution_generator/Select.test]: @zuyu I found that the code logic here is a little bit wrong. For window w/ NULLs, the NULL values should be ignored rather than directly mark the result NULL. Only when all the values in the window are NULLs should the window aggregate result be marked NULL. 

Will fix this in the next commit.

**Comment by shixuan-fan** [comment]
@zuyu I have a new version on my local repo which supports `RANGE`, and also optimizes the calculation from quadratic to linear. Could we have this one merged so that I could open another PR for the new feature? Thanks!

**Comment by zuyu** [comment]
Wait for the CI.

**Comment by shixuan-fan** [comment]
Sounds good. Thanks a lot! :)
